proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,443 @@
+- Start Date: 2023-21-02
+- Reference Issues: #5202 (and all the issues about `@astrojs/image`, really)
+- Implementation PR: https://github.com/withastro/astro/pull/6344
+
+# Summary
+
+This RFC aims to simplify the usage, optimization and resizing of images in Astro.
+
+# Nomenclature
+
+- **CLS**: Content Layout Shift. A score determining how much content shifted on your page during loading, you want to be as close to 0 (or "no CLS") as possible.
+
+# Example
+
+> Note: All the `src` paths in the examples below represent the final paths at build. During development, those paths may be different due to implementations details.
+
+> Note 2: All the results below, unless specified otherwise, are done using the default image services provided with Astro.
+
+## Image Component
+
+### Image located in `src`, in an Astro file
+
+```astro
+---
+import { Image } from 'astro:assets';
+import myImage from "../assets/my_image.png"; // Image is 1600x900
+---
+
+<!-- `alt` is mandatory on the Image component -->
+<Image src={myImage} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/my_image.hash.webp"
+  width="1600"
+  height="900"
+  decoding="async"
+  loading="lazy"
+  alt="..."
+/>
+```
+
+### Remote images
+
+```astro
+---
+import { Image } from "astro:assets";
+// Image is 1920x1080
+---
+
+<Image src="https://example.com/image.png" alt="..." />
+<!-- ERROR! `width` and `height` are required -->
+
+<Image src="https://example.com/image.png" width={1280} alt="..." />
+<!-- ERROR! `height` is required -->
+
+<Image src="https://example.com/image.png" width={1280} height={720} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Remote images are not optimized or resized. Nonetheless, you still benefit from the enforced attributes and the guarantee of no CLS, albeit manually. -->
+<img
+  src="https://example.com/image.png"
+  decoding="async"
+  loading="lazy"
+  width="1280"
+  height="720"
+  alt="..."
+/>
+```
```

Comment:
Do you think pixel-density is within scope of this API?

The output would look something like this:
```html
<img
  srcset="
    /_astro/my_image.dp1hash.webp 1600w,
    /_astro/my_image.dp2hash.webp 3200w
  "
  sizes="(min-width: 3200px) 3200px, 100vw"
  src="/_astro/my_image.dp1hash.webp"
/>
```

I'm not quite sure what the inputs should be. Currently the `@astrojs/image` `Picture` component accepts a `widths: number[]`. But `sizes` and `srcset` are available on `<img>` without a `<picture>`, so I would expect `Image` to support them as well.

At a minimum, I think there should be _some way_ to tell `Image` to generate image files for a list of widths.

Could it be calculated from `srcset`? From `sizes`? Is `widths` required? I'm not sure about these details.

See also [Responsive images with srcset and sizes](https://medium.com/@woutervanderzee/responsive-images-with-srcset-and-sizes-fc434845e948) and [Optimal Images in HTML](https://www.builder.io/blog/fast-images).

Reply:
This is considered out of scope for this current proposal (see the non-goals section, it's part of it)

You can achieve this with the APIs shipped in 2.1 technically, but in the future we aim to have official support, just not in this version.

Reply:
> This is considered out of scope for this current proposal

As long as the APIs proposed here don't _preclude_ it being added in the future, I'm hugely supportive of shipping one thing at a time 😄👍